compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/AssistedFactoryCodeGen.kt

```diff
@@ -559,19 +559,24 @@ object AssistedFactoryCodeGen : AnvilApplicabilityChecker {
             .build(),
         )
         .apply {
+          fun createFactory(name: String, providerTypeName: ClassName): FunSpec {
+            return FunSpec.builder(name)
+              .jvmStatic()
+              .addTypeVariables(typeParameters)
+              .addParameter(DELEGATE_FACTORY_NAME, generatedFactoryTypeName)
+              .returns(providerTypeName.parameterizedBy(baseFactoryTypeName))
+              .addStatement(
+                "return %T.create(%T($DELEGATE_FACTORY_NAME))",
+                InstanceFactory::class,
+                implParameterizedTypeName,
+              )
+              .build()
+          }
           TypeSpec.companionObjectBuilder()
+            .addFunction(createFactory("create", Provider::class.asClassName()))
+            // New in Dagger 2.50: factories for dagger.internal.Provider
             .addFunction(
-              FunSpec.builder("create")
-                .jvmStatic()
-                .addTypeVariables(typeParameters)
-                .addParameter(DELEGATE_FACTORY_NAME, generatedFactoryTypeName)
-                .returns(Provider::class.asClassName().parameterizedBy(baseFactoryTypeName))
-                .addStatement(
-                  "return %T.create(%T($DELEGATE_FACTORY_NAME))",
-                  InstanceFactory::class,
-                  implParameterizedTypeName,
-                )
-                .build(),
+              createFactory("createFactoryProvider", dagger.internal.Provider::class.asClassName()),
```

Comment:
Does Dagger always generate both now? Or are they keying off anything to decide on one vs the other? Looks like all the tests passed so it should be fine, but it feels a little surprising

Reply:
It always generates both now. The original for backward compatibility, the new one for forward compatibility. They're going to support both going forward